### PR TITLE
🚧 [desktop-distribution] Qualify authenticated macOS replacement [step 8/14]

### PR DESCRIPTION
**Pull request by AlexDochioiu**
## Complexity

🚧 Complex: credential-bearing native lifecycle qualification across two signed package versions, classic Keychain ACLs, the real tray Quit path, and two native Mac CPUs.

The final diff is 1,870 changed lines (1,847 additions, 23 deletions), all authored with no generated code. It is 370 lines above the ~1,500 soft target after review required a 158-line native Security.framework helper plus PID-generation, diagnostic, exception-preservation, and regression coverage. The scope remains one coherent private qualification seam; production app code is unchanged.

## What

- Add a separate main-only `macos-authenticated-upgrade-probe` mode behind a dedicated `desktop-qa` environment.
- Serialize x64 and arm64 jobs so one approved QA account never competes for its relay bridge slot.
- Consume and remove credential environment values before child processes and request phase-fresh sessions in memory.
- Compile a probe-only Security.framework writer that receives Keychain values only through stdin and applies ACLs for the installed signed app, `/usr/bin/security`, and itself; no secret enters argv.
- Install the retained N package, persist Bridge On, bind profile/relay readiness markers to one exact packaged-helper PID, invoke the real tray Quit, replace with N+1, and repeat with a fresh bounded session.
- Verify the same authenticated helper remains alive through visible-window readiness, plus Keychain/session, On intent, state/login sentinels, helper shutdown, and no relaunch/orphan.
- Exclude raw auth responses, token values, bridge/app logs, and authenticated screenshots from artifacts while retaining safe native HTTP/Keychain diagnostics and primary+cleanup failures.
- Add focused security/lifecycle regression tests and update the active plan/regression boundaries.

## Why

Private helper-Off replacement is accepted on both Mac CPUs, but authenticated helper-On and Keychain restoration remain open. This creates a reviewable credential boundary without widening `macos-signing`, publication permissions, or the already accepted credential-free probe.

## Risk and test focus

- No QA credential is added or used by this PR. Current local development credentials are explicitly ineligible for the production endpoint.
- Before any dispatch, `desktop-qa` must be configured for exact `main`, and the owner must securely provision an approved production-compatible QA account.
- Secret values are step-scoped, removed from the Python environment before subprocesses, sent only over stdin to the native writer/auth request, and never copied to artifacts.
- The native writer intentionally uses classic `SecKeychain` APIs to match the production app's established non-data-protection Keychain mode; the remaining risk is real ACL behavior on fresh signed-package runners.
- Only a merged-main both-CPU run with approved credentials can establish authenticated evidence.
- Signing, notarization, publication, public downloads, failed-stop behavior, interactive browser/TCC, and minimum-OS gates remain unchanged and open.

## Expected result

After review and merge, an approved credential can drive private authenticated `1.8.4+24 → 1.9.0+62` replacement on native x64 and arm64 without exposing credentials or weakening signing/publication gates. This PR itself has no user-visible or database impact and does not claim authenticated qualification has passed.

## Verification

- `xcrun swiftc .github/scripts/write_desktop_macos_keychain.swift -framework Security -o <temporary-output>`
- `python3 -m py_compile .github/scripts/qualify_desktop_macos_authenticated_upgrade.py .github/scripts/test_qualify_desktop_macos_authenticated_upgrade.py`
- `python3 -m unittest discover -s .github/scripts -p 'test_*desktop*.py'` — 103 passed
- `actionlint .github/workflows/desktop-qualification.yml`
- `git diff --cached --check`

### CODE REVIEW

File: .github/workflows/desktop-qualification.yml [hunk 1]
```diff
@@ -325,6 +328,118 @@ jobs:
...
+          python .github/scripts/qualify_desktop_macos_authenticated_upgrade.py \
+            --previous-run-json build/desktop-macos-authenticated-upgrade/previous/run.json \
+            --previous-packages build/desktop-macos-authenticated-upgrade/previous/packages \
+            --previous-evidence build/desktop-macos-authenticated-upgrade/previous/evidence \
+            --previous-pulls-json build/desktop-macos-authenticated-upgrade/previous/pulls.json \
+            --current-run-json build/desktop-macos-authenticated-upgrade/current/run.json \
+            --current-packages build/desktop-macos-authenticated-upgrade/current/packages \
+            --current-evidence build/desktop-macos-authenticated-upgrade/current/evidence \
+            --current-pulls-json build/desktop-macos-authenticated-upgrade/current/pulls.json \
+            --architecture '${{ matrix.architecture }}' --channel '${{ inputs.channel }}' \
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/workflows/desktop-qualification.yml]: [Sesori reply] Fixed in 59df1fa: the choice input now enters the step through CHANNEL and the shell passes --channel "$CHANNEL"; no workflow expression is interpolated into that QA-secret command.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+
+AUTH_ENDPOINT = "https://api.sesori.com/auth/email"
+AUTH_EMAIL_ENVIRONMENT_KEY = "SESORI_DESKTOP_QA_EMAIL"
+AUTH_PASSWORD_ENVIRONMENT_KEY = "SESORI_DESKTOP_QA_PASSWORD"
+KEYCHAIN_SERVICE = "com.sesori.desktop"
+KEYCHAIN_ACCOUNTS = ("access_token", "refresh_token", "auth_user")
+BRIDGE_LOG = SUPPORT_ROOT / "logs/bridge.log"
+HELPER_WAIT_SECONDS = 60.0
+HELPER_POLL_INTERVAL_SECONDS = 1.0
+MIN_SESSION_VALIDITY_SECONDS = 600
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py:51]: [Sesori reply] Fixed in 59df1fa: the probe now requests a phase-fresh session after the prior launch, updates all three Keychain values before replacing the old signed app, and rechecks a 180-second bounded-launch budget after current-package installation. The report records phase-fresh restoration rather than claiming the original token bytes survived both phases.

File: .plan/active/desktop-distribution/TRACKER.md [hunk 3]
```diff
@@ -257,7 +264,7 @@ notification authorization/delivery, interactive TCC, OS-login and ship gates st
 | Gate | State | Evidence still required |
 |---|---|---|
 | Native build matrix | All six staging rows passed in final 3.a run 34987193233 | Signed/interactive release gates remain unverified. |
-| macOS update path | Private helper-Off passed both CPUs | Authenticated helper-On and public gates pending. |
+| macOS update path | Helper-Off passed; helper-On tooling in progress | Helper-On run, failed-stop and public gates pending. |
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .plan/active/desktop-distribution/TRACKER.md]: [Sesori reply] Fixed in 59df1fa: the modified macOS tracker row is now 107 characters.

File: .plan/active/desktop-distribution/PLAN.md [hunk 4]
```diff
@@ -447,8 +448,29 @@ Evidence artifacts `10574093842` (x64, digest
...
 `NO_INSTALLED_HELPER`, and every implemented `upgrade.json` check is true, including
-`helperAbsentBeforeQuit`. Private helper-Off replacement is accepted on both CPUs;
-authenticated helper-On, failed-stop and the other unchanged macOS gates remain open.
+`helperAbsentBeforeQuit`. Private helper-Off replacement is accepted on both CPUs.
+PR #1547 recorded that accepted boundary and merged as
+`1ff3780b0b9244f5dada842c9e7735a28712fb2d`.
+
+**Step 6 authenticated helper-On qualification PR:**
+`🚧 [desktop-distribution] Qualify authenticated macOS replacement [step 8/14]`.
+Add a separate `macos-authenticated-upgrade-probe` job rather than widening the
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .plan/active/desktop-distribution/PLAN.md]: [Sesori reply] Fixed in 59df1fa: the plan now distinguishes the macos-authenticated-upgrade-probe dispatch mode from the macos-authenticated-upgrade job key.

File: .github/scripts/test_qualify_desktop_macos_authenticated_upgrade.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        source = inspect.getsource(launch_authenticated_and_quit)
+        observation = source.index("wait_for_authenticated_helper(")
+        tray_quit = source.index("quit_result = subprocess.run")
+        self.assertLess(observation, tray_quit)
+        self.assertIn("application/helper remained or relaunched after Quit", source)
+        self.assertIn("private_app_log = SUPPORT_ROOT", source)
+        self.assertNotIn("screencapture", source)
+
+    def test_qualification_seeds_on_before_previous_launch_and_reuses_keychain(self):
+        source = inspect.getsource(qualify)
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/test_qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in 59df1fa: the ordering tests now parse the functions with ast and select exact call objects, assignments, labels, and keyword arguments instead of comparing source substrings.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+                "add-generic-password",
+                "-a",
+                account,
+                "-s",
+                KEYCHAIN_SERVICE,
+                "-T",
+                str(executable),
+                "-T",
+                "/usr/bin/security",
+                "-w",
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in 59df1fa by removing security add-generic-password entirely. A compiled Security.framework helper reads each value only from stdin and applies an ACL for the signed app, /usr/bin/security, and the helper; no secret enters argv.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+                "restoration, authenticated profile lookup, relay serving, visible startup, actual tray Quit "
+                "stopping the helper, no relaunch/orphan, and bounded desktop/shared-state preservation. Not "
+                "interactive browser login, failed-stop refusal, user-account/TCC, public-download, or minimum-OS QA."
+            ),
+        }
+        (output / "authenticated-upgrade.json").write_text(json.dumps(report, indent=2) + "\n", encoding="utf-8")
+        print(f"PASS private authenticated macOS/{architecture} {previous.version}+{previous.build_number} "
+              f"→ {current.version}+{current.build_number} manual replacement")
+    finally:
+        cleanup_qualification(created_keychain_accounts=created_keychain_accounts)
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in 59df1fa: qualification_cleanup now preserves the primary failure and combines a simultaneous cleanup failure in an ExceptionGroup. Focused coverage asserts both original exception types and messages remain available.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+            )
+            raise RuntimeError(f"{label}: expected at most one installed helper process")
+        bridge_output = ""
+        if bridge_log.is_file() and bridge_log.stat().st_size >= bridge_log_offset:
+            with bridge_log.open("rb") as stream:
+                stream.seek(bridge_log_offset)
+                bridge_output = stream.read().decode("utf-8", errors="replace")
+        authenticated_profile = "Authenticated as " in bridge_output
+        relay_serving = "Waiting for relay events..." in bridge_output
+        if helper_process_count == 1 and authenticated_profile and relay_serving:
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in 59df1fa: helper monitoring begins as soon as the app PID appears, snapshots the log at each exact helper PID generation, resets markers on PID changes, and returns that PID. After window readiness, the probe refuses Quit unless the same authenticated PID is still the sole installed helper.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+                "-T",
+                str(executable),
+                "-T",
+                "/usr/bin/security",
+                "-w",
+            ],
+            secret_input=value,
+        )
+        if result.returncode != 0:
+            raise RuntimeError(f"Could not create the {account} desktop Keychain item")
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in 59df1fa: Keychain create/update/read/delete failures now retain bounded return-code and stderr diagnostics while excluding stdout. The stdin secret is defensively redacted from native-writer stderr, with focused coverage.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+        data=body,
+        headers={"Content-Type": "application/json"},
+        method="POST",
+    )
+    try:
+        with urllib.request.urlopen(request, timeout=20) as response:
+            status = response.status
+            payload = json.load(response)
+    except urllib.error.HTTPError as error:
+        raise RuntimeError(f"QA email login returned HTTP {error.code}") from None
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in 59df1fa: the HTTPError is now chained as the RuntimeError cause, preserving its native traceback and metadata without including the request body in the wrapper message.

**Comment by AlexDochioiu** [comment]
[Sesori reply] Inspected the completed Codex review summary. Its five inline findings are fixed in 59df1fa with per-thread replies: native stdin-only Keychain writes, preserved primary+cleanup failures, PID-bound readiness, retained safe diagnostics, and chained HTTP errors.

**Review by AlexDochioiu — COMMENTED** [review]
[Sesori reply] Addressed all ten actionable review findings in 59df1fa, replied on each thread, and resolved them. Verification now includes the Security.framework writer compile, Python compile, 101 desktop tooling tests, actionlint, and diff checks.

File: .github/scripts/write_desktop_macos_keychain.swift [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+guard let operation = Operation(rawValue: CommandLine.arguments[1]) else {
+    fail("Unknown Keychain writer operation")
+}
+
+let account = CommandLine.arguments[2]
+let service = CommandLine.arguments[3]
+let trustedPaths = [
+    CommandLine.arguments[4],
+    CommandLine.arguments[5],
+    URL(fileURLWithPath: CommandLine.arguments[0]).standardizedFileURL.path,
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/write_desktop_macos_keychain.swift]: [Sesori reply] Fixed in aac88e7: a relative argv[0] is now joined to FileManager.default.currentDirectoryPath and then standardized before SecTrustedApplicationCreateFromPath receives it. The Swift helper compiles successfully.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 12]
```diff
@@ -0,0 +1,773 @@
...
+        failures.append(f"remaining owned processes:\n{remaining}")
+    if failures:
+        raise RuntimeError("Authenticated upgrade cleanup failed:\n" + "\n".join(failures))
+
+
+@contextmanager
+def qualification_cleanup(*, cleanup: Callable[[], None]) -> Iterator[None]:
+    try:
+        yield
+    except Exception as qualification_error:
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in aac88e7: qualification_cleanup now catches BaseException, always attempts cleanup for KeyboardInterrupt/SystemExit as well as ordinary failures, and raises BaseExceptionGroup when cleanup also fails. Focused coverage exercises the interrupt-plus-cleanup-failure case.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 13]
```diff
@@ -0,0 +1,773 @@
...
+                label=label,
+                output=output,
+                helper_process_count=helper_process_count,
+                helper_pid=None,
+                authenticated_profile=False,
+                relay_serving=False,
+            )
+            raise RuntimeError(f"{label}: expected at most one installed helper process")
+        if helper_pid != active_helper_pid:
+            generation_log_offset = bridge_log.stat().st_size if bridge_log.is_file() else 0
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in aac88e7: launch_authenticated_and_quit snapshots the bridge-log offset immediately before Popen and passes it into helper observation. The first observed helper retains that pre-launch baseline; only a subsequently observed helper generation resets the offset. Tests cover markers emitted before the first PID poll and the baseline-before-launch ordering.

File: .github/scripts/test_qualify_desktop_macos_authenticated_upgrade.py [hunk 14]
```diff
@@ -0,0 +1,594 @@
...
+
+            recorded = json.loads((root / "current-authenticated-helper.json").read_text(encoding="utf-8"))
+            self.assertFalse(recorded["authenticatedProfileConfirmed"])
+            self.assertFalse(recorded["relayServing"])
+
+    def test_helper_restart_requires_fresh_markers_from_the_new_pid(self):
+        with tempfile.TemporaryDirectory() as directory:
+            root = Path(directory)
+            bridge_log = root / "bridge.log"
+            bridge_log.write_text("Authenticated as first-helper\n", encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/test_qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in aac88e7: the restart test now appends both stale readiness markers for helper 501 before observing helper 502, then asserts a second sleep/new-marker append was required. Removing the generation re-anchor would return early and fail the test.

File: .github/scripts/test_qualify_desktop_macos_authenticated_upgrade.py [hunk 15]
```diff
@@ -0,0 +1,594 @@
...
+            recorded = json.loads((root / "current-authenticated-helper.json").read_text(encoding="utf-8"))
+            self.assertEqual(recorded["helperPid"], 502)
+            self.assertTrue(recorded["authenticatedProfileConfirmed"])
+            self.assertTrue(recorded["relayServing"])
+
+    def test_authenticated_helper_is_observed_and_remains_same_process_before_real_tray_quit(self):
+        events = call_events(launch_authenticated_and_quit)
+        observation = find_event(events=events, name="wait_for_authenticated_helper")
+        window = find_event(events=events, name="wait_for_visible_window")
+        current_process = find_event(events=events, name="_helper_pid")
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/test_qualify_desktop_macos_authenticated_upgrade.py:563]: [Sesori reply] Fixed in aac88e7 with a stronger AST assertion: the exact require condition must remain helper_process_count == 1 and current_helper_pid == authenticated_helper_pid, in addition to the established observation/window/Quit ordering.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 16]
```diff
@@ -0,0 +1,773 @@
...
+            active_helper_pid = helper_pid
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py:393]: [Sesori reply] Fixed in aac88e7: the bridge-log baseline is captured before app launch and retained for the first observed helper PID, so markers emitted before the first process poll are accepted. The baseline resets only after an already-observed helper generation changes.

**Review by AlexDochioiu — COMMENTED** [review]
[Sesori reply] Addressed all six second-wave inline findings in aac88e7, replied on each thread, and resolved them. The follow-up preserves first-generation pre-poll markers, re-anchors replacements, cleans up across BaseException paths, normalizes the writer path, and strengthens exact regression guards. Verification: Swift compile, Python compile, 102 desktop tooling tests, actionlint, and diff checks.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 17]
```diff
@@ -0,0 +1,780 @@
...
+    *,
+    label: str,
+    inspector: Path,
+    quitter: Path,
+    output: Path,
+) -> None:
+    launcher_log = output / f"{label}-launcher.log"
+    # Authenticated app output stays in the exact probe-owned support root and is never uploaded.
+    private_app_log = SUPPORT_ROOT / "desktop-instance" / f"{label}-authenticated-app.log"
+    bridge_log_offset = BRIDGE_LOG.stat().st_size if BRIDGE_LOG.is_file() else 0
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in 60c92dc: each helper poll now compares the active bridge.log size with the generation offset and re-anchors to byte zero when the active file is smaller, then evaluates the rotated file immediately. A focused rotation regression uses an oversized pre-rotation baseline and confirms the ready helper is accepted.

**Review by AlexDochioiu — COMMENTED** [review]
[Sesori reply] Addressed the bridge-log rotation finding in 60c92dc, replied on the thread, and resolved it. Verification: Python compile, 23 focused tests, 103 desktop tooling tests, and diff checks.

File: .github/scripts/qualify_desktop_macos_authenticated_upgrade.py [hunk 18]
```diff
@@ -0,0 +1,781 @@
...
+        if helper_pid != active_helper_pid:
+            if observed_helper_generation:
+                generation_log_offset = bridge_log.stat().st_size if bridge_log.is_file() else 0
+            active_helper_pid = helper_pid
+            if helper_pid is not None:
+                observed_helper_generation = True
+        bridge_output = ""
+        if active_helper_pid is not None and bridge_log.is_file():
+            bridge_log_size = bridge_log.stat().st_size
+            if bridge_log_size < generation_log_offset:
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply .github/scripts/qualify_desktop_macos_authenticated_upgrade.py]: [Sesori reply] Fixed in 7a00bec: the baseline now carries device, inode, and offset. Each poll opens bridge.log first, fstats that exact descriptor, and resets to byte zero when device/inode changes or the same file shrinks. The rotation regression now atomically replaces the log with a different-inode file that is larger than the old offset, covering this case.

**Review by AlexDochioiu — COMMENTED** [review]
[Sesori reply] Addressed the file-identity rotation finding in 7a00bec, replied on the thread, and resolved it. The reader now compares device/inode on the opened descriptor and retains the same-file truncation guard. Verification: Python compile, 23 focused tests, 103 desktop tooling tests, and diff checks.

**Review by AlexDochioiu — COMMENTED** [review]
[Sesori reply] Reviewed the latest approval summary. No additional findings remain; all 18 review threads are resolved, CI passes 21/21, and local verification remains green.
